backend/http_settings.go

```diff
@@ -0,0 +1,249 @@
+	Access            string `json:"access"`
+	URL               string `json:"url"`
+	BasicAuthEnabled  bool   `json:"basicAuth"`
+	BasicAuthUser     string `json:"basicAuthUser"`
+	BasicAuthPassword string `json:"secure.basicAuthPassword"`
+	Headers           map[string]string
+
+	Timeout               time.Duration `json:"timeout"`
+	KeepAlive             time.Duration `json:"httpKeepAlive"`
+	TLSHandshakeTimeout   time.Duration `json:"httpTLSHandshakeTimeout"`
+	ExpectContinueTimeout time.Duration `json:"httpExpectContinueTimeout"`
+	MaxIdleConns          int           `json:"httpMaxIdleConns"`
+	IdleConnTimeout       time.Duration `json:"httpIdleConnTimeout"`
+
+	TLSClientAuth     bool   `json:"tlsAuth"`
+	TLSAuthWithCACert bool   `json:"tlsAuthWithCACert"`
+	TLSSkipVerify     bool   `json:"tlsSkipVerify"`
+	TLSServerName     string `json:"serverName"`
+	TLSCACert         string `json:"secure.tlsCACert"`
+	TLSClientCert     string `json:"secure.tlsClientCert"`
+	TLSClientKey      string `json:"secure.tlsClientKey"`
+
+	SigV4Auth          bool   `json:"sigV4Auth"`
+	SigV4Region        string `json:"sigV4Region"`
+	SigV4AssumeRoleARN string `json:"sigV4AssumeRoleArn"`
+	SigV4AuthType      string `json:"sigV4AuthType"`
+	SigV4ExternalID    string `json:"sigV4ExternalId"`
+	SigV4Profile       string `json:"sigV4Profile"`
+	SigV4AccessKey     string `json:"secure.sigV4AccessKey"`
+	SigV4SecretKey     string `json:"secure.sigV4SecretKey"`
+}
+
+// HTTPClientOptions creates and returns httpclient.Options.
+func (s *HTTPSettings) HTTPClientOptions() *httpclient.Options {
+	opts := httpclient.Options{
+		Headers: s.Headers,
+	}
+
+	opts.Timeouts = &httpclient.TimeoutOptions{
+		Timeout:               s.Timeout,
+		KeepAlive:             s.KeepAlive,
+		TLSHandshakeTimeout:   s.TLSHandshakeTimeout,
+		ExpectContinueTimeout: s.ExpectContinueTimeout,
+		MaxIdleConns:          s.MaxIdleConns,
+		IdleConnTimeout:       s.IdleConnTimeout,
+	}
+
+	if s.BasicAuthEnabled {
+		opts.BasicAuth = &httpclient.BasicAuthOptions{
+			User:     s.BasicAuthUser,
+			Password: s.BasicAuthPassword,
+		}
+	}
+
+	if s.TLSClientAuth || s.TLSAuthWithCACert {
+		opts.TLS = &httpclient.TLSOptions{
+			CACertificate:      s.TLSCACert,
+			ClientCertificate:  s.TLSClientCert,
+			ClientKey:          s.TLSClientKey,
+			InsecureSkipVerify: s.TLSSkipVerify,
+			ServerName:         s.TLSServerName,
+		}
+	}
+
+	if s.SigV4Auth {
+		opts.SigV4 = &httpclient.SigV4Config{
+			AuthType:      s.SigV4AuthType,
+			Profile:       s.SigV4Profile,
+			AccessKey:     s.SigV4AccessKey,
+			SecretKey:     s.SigV4SecretKey,
+			AssumeRoleARN: s.SigV4AssumeRoleARN,
+			ExternalID:    s.SigV4ExternalID,
+			Region:        s.SigV4Region,
+		}
+	}
+
+	return &opts
+}
+
+//gocyclo:ignore
+func parseHTTPSettings(jsonData json.RawMessage, secureJSONData map[string]string) (*HTTPSettings, error) {
+	s := &HTTPSettings{
+		Headers: map[string]string{},
+	}
+
+	var dat map[string]interface{}
```

Comment:
Is it possible to avoid using map[string]interface{} here and unmarshal into a struct?

Reply:
It's not possible to unmarshal both JSONData and SecureJSONData easily to struct why I choose this path. The idea when creating a HTTP client is that it will be used until (app/data source instance) settings changes so this should not happen that often. In combination with instance management, see [example](https://pkg.go.dev/github.com/grafana/grafana-plugin-sdk-go@v0.92.0/backend/datasource#example-package).

But good point. Might try to use some temporary structs instead than `map[string]interface{}`

Reply:
I mostly thought about readability and avoiding type conversions without `ok` checks, not about performance. But - I just noticed that this unmarshaling involves parsing non-deterministic number of headers, not sure it's possible to do with one struct, do you mean using several unmarshals instead of one to `map[string]interface{}` ?

Reply:
Yeah the unmarshaling is a bit tricky actually so I'll skip trying to unmarshal to a struct.